File: solutions/PHP/findTarget.php
<?php

/****
PHP class to perform binary search on an array.

Author: Uthman Dev

Description: This class contains a method to perform binary search on a sorted array.
             It returns the index of the target if found, or -1 if not found.
****/

class BinarySearch
{
    private array $arr; // Array in which to search
    private int $target; // Target value to search for

    // Constructor to initialize the array and target
    public function __construct(array $arr, int $target)
    {
        $this->arr = $arr;
        $this->target = $target;
    }

    // Method to perform binary search
    public function search(): int // Return type is an integer (index or -1 if not found)
    {
        $start = 0; // Starting index of the search range
        $end = count($this->arr) - 1; // Ending index of the search range

        // Loop to continue the search as long as start index is less than or equal to end index
        while ($start <= $end) {
            $mid = intdiv($start + $end, 2); // Calculate the middle index (floor division)

            // Check if the middle element is less than the target
            if ($this->arr[$mid] < $this->target) {
                $start = $mid + 1; // Move the start index right if the target is larger
            } 
            // Check if the middle element is greater than the target
            elseif ($this->arr[$mid] > $this->target) {
                $end = $mid - 1; // Move the end index left if the target is smaller
            } 
            // If the middle element equals the target, return the index
            else {
                return $mid;
            }
        }
        
        // Return -1 if target is not found in the array
        return -1;
    }
}

// Example usage:
$target = 7;
$n = 100000;
$arr = range(1, $n); // Generate an array of numbers from 1 to 100000
$binarySearch = new BinarySearch($arr, $target);
print_r($binarySearch->search()); // Output: 6

?>

File: solutions/PYTHON/findTarget.py
# Python function to perform binary search on a sorted array.
#
# Author: Uthman Dev
#
# Description: This function takes a sorted array and a target value, and performs a binary
#              search to find the index of the target. If the target is not found, it returns -1.
#
# Parameters:
# arr (list): A sorted list of integers.
# target (int): The integer value to search for.
#
# Returns:
# int: The index of the target in the array, or -1 if the target is not found.

def binary_search(arr, target):
    start = 0  # Initialize the start pointer at the beginning of the array
    end = len(arr) - 1  # Initialize the end pointer at the end of the array

    # Perform binary search
    while start <= end:
        mid = (start + end) // 2  # Calculate the middle of the range

        # Debugging output to verify pointer positions
        print(f"Start: {start}, End: {end}, Mid: {mid}, Mid Element: {arr[mid]}")

        # Check if the middle element is the target
        if arr[mid] == target:
            return mid  # Return the index if found

        # Adjust search range based on the middle element's value
        if arr[mid] < target:
            start = mid + 1  # Search the right half
        else:
            end = mid - 1  # Search the left half

    # Return -1 if the target is not found
    return -1

# Example usage
arr = list(range(1, 100001))  # A large sorted array
target = 7  # The target value to find
result = binary_search(arr, target)  # Call the function with the array and target

# Print the result
print("The target index is:", result)  # Expected output: 6 if target = 7